codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/generators/Instantiator.kt

```diff
@@ -76,6 +78,16 @@ open class Instantiator(
         val lowercaseMapKeys: Boolean = false,
     )
 
+    /**
+     * Client and server structures have different builder types. `Instantiator` needs to know how the builder
+     * type behaves to generate code for it.
+     */
+    interface BuilderKindBehavior {
+        fun hasFallibleBuilder(shape: StructureShape): Boolean
+        fun setterName(memberShape: MemberShape): String
+        fun doesSetterTakeInOption(memberShape: MemberShape): Boolean
```

Comment:
I think these deserve some comments since builders (at least for the client) have two setter methods per field, one of which takes an option if it's an optional field.